x-pack/test/api_integration/deployment_agnostic/apis/observability/apm/settings/agent_keys/agent_keys.spec.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import expect from '@kbn/expect';
+import { PrivilegeType, ClusterPrivilegeType } from '@kbn/apm-plugin/common/privilege_type';
+import { RoleCredentials } from '../../../../../services';
+import { DeploymentAgnosticFtrProviderContext } from '../../../../../ftr_provider_context';
+import { expectToReject } from '../../../../../../../apm_api_integration/common/utils/expect_to_reject';
+import { ApmApiError } from '../../../../../services/apm_api';
+
+export default function ApiTest({ getService }: DeploymentAgnosticFtrProviderContext) {
+  const apmApiClient = getService('apmApi');
+  const samlAuth = getService('samlAuth');
+
+  const agentKeyName = 'test';
+  const allApplicationPrivileges = [PrivilegeType.AGENT_CONFIG, PrivilegeType.EVENT];
+  const clusterPrivileges = [ClusterPrivilegeType.MANAGE_OWN_API_KEY];
+
+  async function createAgentKey(
+    apiClient: (typeof apmApiClient)[keyof typeof apmApiClient],
+    roleAuthc: RoleCredentials,
+    privileges = allApplicationPrivileges
+  ) {
+    return await apiClient({
+      endpoint: 'POST /api/apm/agent_keys 2023-10-31',
+      params: {
+        body: {
+          name: agentKeyName,
+          privileges,
+        },
+      },
+      roleAuthc,
+    });
+  }
+
+  async function invalidateAgentKey(id: string) {
+    return await apmApiClient.writeUser({
+      endpoint: 'POST /internal/apm/api_key/invalidate',
+      params: {
+        body: { id },
+      },
+    });
+  }
+
+  async function getAgentKeys() {
+    return await apmApiClient.writeUser({ endpoint: 'GET /internal/apm/agent_keys' });
+  }
+
+  describe('When the user does not have the required privileges', () => {
+    let roleAuthc: RoleCredentials;
+
+    before(async () => {
+      roleAuthc = await samlAuth.createM2mApiKeyWithRoleScope('editor');
+    });
+
+    describe('When the user does not have the required cluster privileges', () => {
+      it('should return an error when creating an agent key', async () => {
+        const error = await expectToReject<ApmApiError>(() =>
+          createAgentKey(apmApiClient.writeUser, roleAuthc)
```

Comment:
The `POST /api/apm/agent_keys 2023-10-31` endpoint, when used with the `apmApiClient`, specifically requires the `roleAuthc` parameter.

From what I observed in the code, this is necessary because the endpoint is not considered internal, so additional credential information must be included in the request headers.

It seems the credentials require an API key, but the `apmApiClient` does not have one by default. To address this, we need to create a new role with the appropriate API key attached, ensuring the credentials are correctly passed in the request headers.

I’m not sure about the reasoning behind handling this outside of the `apmApiClient` instead of within it. Maybe it’s related to the need to invalidate the API key after the tests are run, avoiding the creation and invalidation of keys for every request that requires one.  

Perhaps @crespocarlos could share more details about why it’s done this way in case you want to know more about it.

Reply:
Now that I'm looking at it, it is weird. We need to improve it.

`roleAuthc` needs to be passed to `apmApi` because, as you've said, we need to destroy the key after the test runs, but calling `apmApi` service with `writeUser` and then having to pass the apiKey looks very strange.

Reply:
For now, lets not use the `apmApi` to call public APIs. It's a momentary compromise on the strongly typed response, but it's less confusing. Let's use the `SupertestWithRoleScopeType` in this case.

we could type the response manually as:

```ts
let response: APIReturnType<'POST /api/apm/agent_keys 2023-10-31'>
```

Reply:
Suggestion: https://github.com/iblancof/kibana/pull/1

Reply:
I’ll be using the new `apmApiClient.publicApi` introduced in https://github.com/iblancof/kibana/pull/1.

https://github.com/elastic/kibana/pull/200762/commits/d50e6bfc6751500f04cb09e4421640112c2bd911